analytics.ts

```diff
@@ -0,0 +1,121 @@
+import * as Settings from '@taqueria/protocol/Settings';
+import { chain, map, resolve } from 'fluture';
+import { pipe } from 'https://deno.land/x/fun@v1.0.0/fns.ts';
+import { getMachineId } from 'https://deno.land/x/machine_id@v0.3.0/mod.ts';
+import type { DenoArgs } from './taqueria-types.ts';
+import * as utils from './taqueria-utils/taqueria-utils.ts';
+
+const {
+	doesPathExist,
+	readJsonFile,
+	writeJsonFile,
+	eager,
+} = utils.inject({
+	stdout: Deno.stdout,
+	stderr: Deno.stderr,
+});
+
+const consentPrompt =
+	'Allow Taqueria anonymously report usage statistics, in accordance with the privacy policy to help make Taqueria better? [y/yes] or [n/no]';
+const optInConfirmationPrompt =
+	'Allow Taqueria anonymously report usage statistics, in accordance with the privacy policy to help make Taqueria better? [y/yes] or [n/no]';
+const optOutConfirmationPrompt = 'Are you sure you want to turn off usage statistic reporting? [y/yes] or [n/no]';
+export const settingsFilePath = Deno.env.get('HOME') + '/.taq-settings.json';
+export const OPT_IN = 'opt_in';
+export const OPT_OUT = 'opt_out';
+
+export const optInAnalyticsFirstTime = () => createSettingsFileWithConsent(OPT_IN);
+
+export const optOutAnalyticsFirstTime = () => createSettingsFileWithConsent(OPT_OUT);
+
+const createSettingsFileWithConsent = (option: 'opt_in' | 'opt_out') =>
+	pipe(
+		Settings.make({ consent: option }),
+		chain(writeJsonFile(settingsFilePath)),
+	);
+
+export const optInAnalytics = () => writeConsentValueToSettings(OPT_IN);
+
+export const optOutAnalytics = () => writeConsentValueToSettings(OPT_OUT);
+
+const writeConsentValueToSettings = (option: 'opt_in' | 'opt_out') => {
+	const input = prompt(option === OPT_IN ? optInConfirmationPrompt : optOutConfirmationPrompt);
+	if (input && /^y(es)?$/i.test(input)) {
+		return pipe(
+			readJsonFile<Settings.t>(settingsFilePath),
+			map((settingsContent: Settings.t) => {
+				settingsContent.consent = option;
+				return settingsContent;
+			}),
+			chain(writeJsonFile(settingsFilePath)),
```

Comment:
Not for this PR, but there is a chance of concurrency problem in this code, because the settings file is not locked between the read and write, another code can read/write it in between, and after both make their changes, one change can be lost
@mweichert  This is very unlikely to happen now, but as the usage patterns get more complex, it can happen. Am I right about the possibility?

Reply:
Exactly, doing that clearly is not something trivial, especially as we are not writing stand-alone code, but code that lives inside VsCode.
The issue: #848